modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/seedlist/ranges.scala

```diff
@@ -0,0 +1,67 @@
+package org.tessellation.node.shared.domain.seedlist
+
+import cats.syntax.all._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.node.shared.domain.seedlist.Range.ValidationResult
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.NonNegLong
+
+case class InvalidInput(str: String) extends NoStackTrace {
+  override def getMessage = s"Invalid input: $str"
+}
+
+object ranges {
+
+  private val delimiter = ";"
+
+  def decode(input: String): ValidationResult[List[Range]] =
+    input
+      .split(delimiter)
+      .toList
+      .traverse(Range.decode)
+      .andThen(validate)
+      .map(mergeOverlaps)
+
+  private def mergeOverlaps(ranges: List[Range]): List[Range] =
+    ranges
+      .sortBy(_.start.value)
+      .foldLeft(List.empty[Range]) {
+        case (prev :: tail, curr) if curr.start - prev.end <= 1 =>
+          val merged = Range(
+            prev.start,
+            NonNegLong.unsafeFrom(
+              Math.max(prev.end, curr.end)
+            )
+          )
+
+          merged :: tail
+        case (acc, curr) =>
+          curr :: acc
+      }
+      .reverse
+
+  private def validate(ranges: List[Range]): ValidationResult[List[Range]] = {
+    val filtered = ranges.filter(r => r.start <= r.end)
+    val input = ranges.foldLeft("") {
+      case (acc, curr) =>
+        val str = s"${curr.start}-${curr.end}"
+
+        if (acc.nonEmpty)
+          acc + ";" + str
+        else
+          str
+    }
+
+    Either
+      .cond(
+        filtered.nonEmpty,
+        filtered,
+        InvalidInput(input)
+      )
+      .toValidatedNel
+  }
```

Comment:
A few thoughts on this new function:

1. From the resulting `Either` it looks like some empty ranges are ok and we won't report them unless they're all empty. Just confirming that's the behavior you want?
2. We're constructing an error message -- iterating over all the ranges -- even though we may not use it
3. I don't see constructing that message as part of this function; IMHO the caller somewhere higher up can do that with something like `invalidRanges.map(_.show).mkString(";")` then you can log that, or return it in an error object, etc. Or maybe skip that step entirely and return the original input string, since it looks like this will only be flagged if all the given ranges are empty.
4. Is there still a benefit in returning `Validated` from this function if you're collapsing the invalid inputs to a single error object?
5. Since it looks like validation of a range is independent of the other ranges around it we can probably simplify things greatly by doing something like `val (invalid, valid) = ranges.partition(_.isEmpty)` and then given those two lists decide under what conditions we are returning, say, _Left(invalid)_ or _Right(valid)_.

Reply:
I agree with above. I have one more thought - shouldn't it return `ValidationResult[NonEmptyList[Range]]` then? as as I see, your main goal here is to filter out empty ranges, so the valid result should always be non empty 🤔

Reply:
1. That's right. As long as there is at least one valid range, the invalid ranges are not reported. That's my intention. Do you have another view point?
2. The error message has been moved to the calling function.

```scala
  def decode(input: String): ValidationResult[List[Range]] =
    input
      .split(delimiter)
      .toList
      .traverse(Range.decode)
      .andThen {
        validate(_)
          .leftMap(_ => InvalidInput(input))
          .toValidatedNel
      }
      .map(mergeOverlaps)
```

3. See the response for point 2.
4. The function has been re-worked to:

```scala
  private def validate(ranges: List[Range]): Either[Throwable, List[Range]] = {
    val (valid, _) = ranges.partition(r => r.start <= r.end)

    Either
      .cond(
        valid.nonEmpty,
        valid,
        NoValidRanges
      )
  }
```

5. See the response for point 4.

Reply:
@cngo-github so basically you've just changed the algorithm. previously we were reporting **all** errors, even if there was at least one correct range. right now we consider it a valid input string if there is at least one correct range but everything is invalid. is that what you wanted?

Reply:
@marcinwadon @sdavidp , my intention is to ignore invalid ranges as long as there is at least one valid range. So, `validate` simply filters out invalid ranges. I would love to report the invalid ranges via logging without stopping the decode process, but I am not sure how to do that given the limitations imposed by `emap` in the `CellDecoder`.

```scala
  implicit val optionalSnapshotOrdinalTimelineCellDecoder: CellDecoder[Option[SnapshotOrdinalTimeline]] =
    CellDecoder.stringDecoder.emap { s =>
      decode(s)
        .leftMap(
          _.map(_.getMessage).toList
            .mkString("; ")
        )
        .leftMap(new DecoderError(_))
        .map(_.some)
        .toEither
    }
```

Per @marcinwadon 's request `validate` now returns a `NonEmptyList[Range]`. So the functions in `ranges` look like this now:

```scala
  def decode(input: String): ValidationResult[NonEmptyList[Range]] =
    input
      .split(delimiter)
      .toList
      .traverse(Range.decode)
      .andThen {
        validate(_)
          .leftMap(_ => InvalidInput(input))
          .toValidatedNel
      }
      .map(mergeOverlaps)

  private def mergeOverlaps(ranges: NonEmptyList[Range]): NonEmptyList[Range] = {
    val merged = ranges
      .sortBy(_.start.value)
      .foldLeft(List.empty[Range]) {
        case (prev :: tail, curr) if curr.start - prev.end <= 1 =>
          val merged = Range(
            prev.start,
            NonNegLong.unsafeFrom(
              Math.max(prev.end, curr.end)
            )
          )

          merged :: tail
        case (acc, curr) =>
          curr :: acc
      }
      .reverse

    NonEmptyList(merged.head, merged.tail)
  }

  private def validate(ranges: List[Range]): Either[Throwable, NonEmptyList[Range]] = {
    val filtered = ranges.filter(r => r.start <= r.end)

    Either
      .cond(
        filtered.nonEmpty,
        NonEmptyList(filtered.head, filtered.tail),
        NoValidRanges
      )
  }
```

and `decode` in `snapshotOrdinalTimeline` is now:

```scala
  def decode(input: String): Validated[NonEmptyList[Throwable], SnapshotOrdinalTimeline] =
    ranges
      .decode(input)
      .map(SnapshotOrdinalTimeline(_))
```

Reply:
I'll take a look in a moment. by the way, you don't need to copy the code, we can see each and every diff in github

Reply:
Ok @cngo-github. I don't understand why would you like to do the following

> my intention is to ignore invalid ranges as long as there is at least one valid range

as in my opinion, if the seedlist has invalid entries, we should not allow for running the node. the seedlist should be fixed and then node restarted. we can ensure seedlist correctness on CI checks. otherwise, you may not figure out you have invalid entries and run the network without some nodes.

Anyway, if you truly believe it's a way to go, as I see in the code, for such an entry `1-15;1 6-20` you would invalidate the whole entry, but for `1-15;20-16` you would alter it to `1-15`. It looks like inconsistency for me.

Reply:
@marcinwadon , you have a point. I've re-worked the code to fail on any invalid `Range`s.